File: courier/vector.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- 
"""
Some aux. classzorz for simple vector arithmatic
"""
import math

def angle(a,b):
    """
    I'm making this a helper function because object-orientation stinks. I like
    to think of it not as one vector comparing its angle to another vector, but
    simply the angle_between(vector1, vector2).
    """
    # Dot product of A, B = |A||B|cos(θ)
    # ⇒ cos(θ) = D ÷ (|A||B|)
    # ⇒ θ = acos(D ÷ (|A||B|))
    return math.acos(a*b / (abs(a.dist) * abs(b.dist)))

def angle_normalize(angle):
    """
    Given an angle in radians, will return an equivalent angle between
    [-pi, pi]
    """
    return (angle + math.pi) % (2*math.pi) - math.pi

def rotate(v, a):
    """Rotate a 2D vector by some angle in radians, a.
    rotate(0, 1, math.radians(90)) will rotate the point (0, 1) 90° about the
    origin and return (-1, 0). Warning: Will ignore the Z component."""
    x, y = v
    return Vector([x*math.cos(a) -y*math.sin(a), 
            x*math.sin(a) + y*math.cos(a)])

def projection(a,b):
    """
    Projection of A onto B. It will have the right length, but B's direction.
    """
    u = b.dir
    u.dist = a*u
    return u

class Vector(list):
    def __add__(self, other):
        return Vector([p+q for p,q in zip(self,other)])

    def __mul__(self, other):
        " Dot product "
        return sum([p*q for p,q in zip(self, other)])

    def __div__(self, other):
        pass
    
    def __sub__(self, other):
        return Vector([p-q for p,q in zip(self,other)])

    def __str__(self):
        return "<%s>" % ", ".join([str(elt) for elt in self])

    def _get_dist(self):
        return math.sqrt(sum([x**2 for x in self]))

    def _set_dist(self, newdist):
        ratio = float(self.dist / newdist)
        for n, x in enumerate(self):
            self[n] = x/ratio
    dist = property(_get_dist, _set_dist)

    @property
    def unitp(self):
        return self.dist == 1

    @property
    def dir(self):
        """ Return a unit vector that points in the same direction """
        u = Vector(self)
        u.dist = 1
        return u

    @property
    def angle(self):
        """ Returns an angle in radians from the positive X axis """
        tmp = angle(Vector([1,0]), self)
        return tmp if self[1] >= 0 else -tmp

if __name__ == "__main__":
    print "------ Vectors ------"
    x = Vector([2,3,4,5])
    y = Vector([1,2,3,4])
    print "x=%s   y=%s" % (x,y)
    print "x+y = %s\nx*y = %s" % (x+y, x*y)
    print "Distance of x: %f" % x.dist

    print "\n\n------ Unit vectors ------"

    print "x is a unit vector" if x.unitp else "x is not a unit vector"
    print "A unit vector of x is %s" % x.dir
    print "x is still %s" % x
    z = Vector([3./5,4./5,0])
    print "%s is a unit vector" % z if z.unitp else "%s is not a unit vector" % z
    x.dist = 2
    print "x now has distance of 2: %s" % x


    print "\n\n------ Angles ------"
    r = Vector([9, 0])
    p = Vector([1, math.sqrt(3)])
    print "Angle between %s and %s is %f" % (r, p, angle(r,p))
    print "(That's %f° btw)" % math.degrees(angle(p,r))

    x,y = Vector([1,1]), Vector([-1,-2])
    print "Angle between %s and %s is %f" % (x,y,math.degrees(angle(x,y)))

    print "\n\n------ Projections ------"
    print p
    print "That vector projected against a 45° going up and to the right:"
    print projection(p, Vector([1,1]))
    print "\n\n------ Rotations ------"
    print Vector([5, 3])
    print "Rotated 45° ccw:"
    print rotate(Vector([5, 3]), math.radians(45))

File: server/physics/vector.js
// Vector manipulation

function Vector(x, y) {
  // Vector. Sorta immutable; none of these methods actually make any permanent
  // changes to the object.
  this.x = x;
  this.y = y;
}

Vector.prototype.add = function(other) {
  // Return a copy of this added to other
  return new Vector(this.x + other.x, this.y + other.y);
};

Vector.prototype.sub = function(other) {
  return new Vector(this.x - other.x, this.y - other.y);
};

Vector.prototype.multiply = function(d) {
  return new Vector(this.x*d, this.y*d);
};

Vector.prototype.dot = function(other) {
  // Returns the dot product of this against the other
  return this.x * other.x + this.y * other.y;
};

Vector.prototype.dist = function() {
  // Returns or sets the vector's distance, and by "set" we really mean "returns
  // a copy of this vector with the new distance"
  if (typeof arguments[0] == 'undefined') {
    return Math.sqrt(this.x*this.x + this.y*this.y);
  } else {
    return this.toUnit().multiply(arguments[0]);
  }
};

Vector.prototype.toUnit = function() {
  // Returns a unit vector of us
  return this.multiply(1/this.dist());
};

Vector.prototype.angleTo = function(other) {
  // Dot product of A, B = |A||
  // => cos(θ) = D ÷ (|A||B|)
  // => θ = acos(D ÷ (|A||B|))
  return Math.acos(this.dot(other) / (Math.abs(this.dist()) * Math.abs(other.dist())));
};

Vector.prototype.copy = function() {
  // Return a copy of this vector
  return new Vector(this.x, this.y);
};

Vector.prototype.rotated = function(a) {
  // Return a new vector that's a copy of this vector rotated by a radians
  return new Vector(this.x * Math.cos(a) - this.y*Math.sin(a),
                    this.x * Math.sin(a) + this.y*Math.cos(a));
};

Vector.prototype.projectedOn = function(other) {
  // Return a copy of ourselves projected to another vector. This vector will
  // have other's direction, but the correct length.
  var u = other.toUnit();
  return u.dist(this, u);
};

function normalizeAngle(angle) {
  // Given an angle in radians, will return an equivalent angle between
  // [-pi, pi]
  return (angle + Math.PI) % (2*Math.PI) - Math.PI;
}


process.mixin(exports,
  {
    Vector: Vector,
    '2DVector': Vector,
    normalizeAngle: normalizeAngle
  }
);
